File: Leetcode-one-by-one/src/Maximize Spanning Tree Stability with Upgrades.go
import (
	"math"
	"sort"
)

func maxStability(n int, edges [][]int, k int) int {

	parent := make([]int, n)
	rank := make([]int, n)

	for i := 0; i < n; i++ {
		parent[i] = i
	}

	var find func(int) int
	find = func(u int) int {
		if parent[u] == u {
			return u
		}
		parent[u] = find(parent[parent[u]])
		return parent[u]
	}

	union := func(u, v int) bool {
		pu := find(u)
		pv := find(v)

		if pu == pv {
			return false
		}

		if rank[pu] >= rank[pv] {
			parent[pv] = pu
			if rank[pu] == rank[pv] {
				rank[pu]++
			}
		} else {
			parent[pu] = pv
		}

		return true
	}

	sort.Slice(edges, func(i, j int) bool {
		if edges[i][3] == edges[j][3] {
			return edges[i][2] > edges[j][2]
		}
		return edges[i][3] > edges[j][3]
	})

	comp := n
	minOne := math.MaxInt32
	minZero := math.MaxInt32

	pq := []int{}

	for _, e := range edges {
		u, v, w, t := e[0], e[1], e[2], e[3]

		ok := union(u, v)

		if t == 1 && !ok {
			return -1
		}

		if ok {
			comp--

			if t == 0 {
				pq = append(pq, w)
			} else {
				if w < minOne {
					minOne = w
				}
			}
		}
	}

	sort.Ints(pq)

	for _, val := range pq {
		if k > 0 {
			if val*2 < minZero {
				minZero = val * 2
			}
			k--
		} else {
			if val < minZero {
				minZero = val
			}
		}
	}

	if comp != 1 {
		return -1
	}

	if minOne < minZero {
		return minOne
	}
	return minZero
}

File: Leetcode-one-by-one/src/Maximize Spanning Tree Stability with Upgrades.swift
import Foundation

class Solution {
    var parent: [Int] = []
    var rank: [Int] = []

    func maxStability(_ n: Int, _ edges: [[Int]], _ k: Int) -> Int {
        parent = Array(0..<n)
        rank = Array(repeating: 0, count: n)

        var edges = edges
        edges.sort {
            if $0[3] == $1[3] {
                return $0[2] > $1[2]
            }
            return $0[3] > $1[3]
        }

        var comp = n
        var minOne = Int(1e9)
        var minZero = Int(1e9)

        var pq: [Int] = []

        for e in edges {
            let didUnion = union(e[0], e[1])

            if e[3] == 1 && !didUnion {
                return -1
            }

            if didUnion {
                comp -= 1

                if e[3] == 0 {
                    pq.append(e[2])
                } else {
                    minOne = min(minOne, e[2])
                }
            }
        }

        pq.sort()

        var k = k
        for val in pq {
            if k > 0 {
                minZero = min(minZero, val * 2)
                k -= 1
            } else {
                minZero = min(minZero, val)
            }
        }

        if comp != 1 { return -1 }
        return min(minOne, minZero)
    }

    func union(_ u: Int, _ v: Int) -> Bool {
        let pu = find(u)
        let pv = find(v)

        if pu == pv { return false }

        if rank[pu] >= rank[pv] {
            parent[pv] = pu
            if rank[pu] == rank[pv] {
                rank[pu] += 1
            }
        } else {
            parent[pu] = pv
        }

        return true
    }

    func find(_ u: Int) -> Int {
        if parent[u] == u { return u }
        parent[u] = find(parent[parent[u]])
        return parent[u]
    }
}
